docs/session_state_api.md

```diff
@@ -0,0 +1,181 @@
+if 'foo' not in st.session_state:
+	st.session_state.foo = 'bar'
+```
+
+### Reads and updates
+
+Read the value of an item in Session State by passing the item to `st.write` :
+
+```python
+# Reads
+st.write(st.session_state.foo)
+
+# Outputs: bar
+```
+
+Update an item in Session State by assigning it a value:
+
+```python
+st.session_state.foo = 'bar2'     # Attribute API
+st.session_state['foo'] = 'bar2'  # Dictionary like API
+```
+
+Curious about what is in Session State? Use `st.write` or magic:
+
+```python
+st.write(st.session_state)
+
+# With magic:
+st.session_state
+```
+
+Streamlit throws a handy exception if an uninitialized variable is accessed:
+
+```python
+st.write(st.session_state['bar'])
+
+# Throws an exception!
+```
+
+![state-uninitialized-exception](_static/img/state_uninitialized_exception.png)
+
+### Delete items
+
+Delete items in Session State using the syntax to delete items in any Python dictionary:
+
+```python
+# Delete a single key-value pair
+del st.session_state[key]
+
+# Delete all the items in Session state
+for key in st.session_state.keys():
+    del st.session_state[key]
+```
+
+Session State can also be cleared by going to Settings → Clear Cache, followed by Rerunning the app.
+
+<img src="_static/img/clear_cache.png" alt="Clear Cache" style='display: block; margin-left: auto; margin-right: auto; padding-top: 15px;'/>
+
+### Session State and Widget State association
+
+Every widget with a key is automatically added to Session State:
+
+```python
+st.text_input("Your name", key="name")
+
+# This exists now:
+st.session_state.name
+```
+
+### Use Callbacks to update Session State
+
+A callback is a python function which gets called when an input widget changes.
+
+**Order of execution**: When updating Session state in response to **events**, a callback function gets executed first, and then the app is executed from top to bottom.
+
+Callbacks can be used with widgets using the parameters `on_change` (or `on_click`), `args`, and `kwargs`:
+
+**Parameters**
+
+- **on_change** or **on_click** - The function name to be used as a callback
+- **args** (_tuple_) - List of arguments to be passed to the callback function
+- **kwargs** (_dict_) - Named arguments to be passed to the callback function
+
+Widgets which support the `on_change` event:
+
+- `st.checkbox`
+- `st.color_picker`
+- `st.date_input`
+- `st.multiselect`
+- `st.number_input`
+- `st.radio`
+- `st.select_slider`
+- `st.selectbox`
+- `st.slider`
+- `st.text_area`
+- `st.text_input`
+- `st.time_input`
+- `st.file_uploader`
+
+Widgets which support the `on_click` event:
+
+- `st.button`
+- `st.form_submit_button`
+
+To add a callback, define a callback function **above** a widget and pass it to the widget via the `on_change` (or `on_click`) parameter.
+
+```eval_rst
+.. note:: A callback is a **function**. In Python, functions are considered *first class*, which means they can be passed as parameters.
```

Comment:
This distinction seems unnecessary in terms of Streamlit docs, since it applies in Python generally.

Reply:
Agreed! Removing the note.